content/guest-post/how-to-authenticate-flask-api-using-jwt/index.md

```diff
@@ -0,0 +1,754 @@
+
+This authentication is used when a user provides a username/email/phone number and a password. This is the most common and weakest authentication factor. The user simply inputs the email and password, and the system checks if the data is valid; if valid, the user gets authenticated and can access the resource. What happens if another person who is not the legitimate user tries to access the resource? The system denies access to the resource.
+
+### [Multi-factor Authentication](https://www.loginradius.com/multi-factor-authentication/)
+
+This authentication uses more than one factor to authenticate a user. For example, the user tries to log in with, say, email and password; if the data is correct, a code is sent to the user's phone number, and the user is asked to input the code. If the user enters the code, the user gets logged in; otherwise, the user is not authenticated. Some applications even go a step further by not using two factors but using three factors.
+
+## Types of Authentication
+
+There are basically three types of authentication, as follows:
+
+1. **Knowledge Authentication**: The user is asked something that only they can provide or know -- e.g., password. This is the most common type and also the easiest.
+2. **Property Authentication**: The user is asked for something they own or possess. For example, they can be asked their pet's name. The idea here is that users will be asked to set a security question and answer when creating an account. This isn’t always used alone; it’s used alongside another authentication type, say, `Knowledge authentication`.
+3. **Biological Authentication**: The user is asked to verify their identity using something biologically unique to them -- e.g, a fingerprint or iris scan.
+
+In most applications, knowledge and property authentication are used as an extra layer of the authentication.
+
+
+## Authentication vs. Authorization
+
+The following are the differences between authentication and authorization:
+
+1. Authentication verifies identity (usually through credential validation)) while authorization grants or denies permissions to a user.
+
+2. Authentication is used to verify that a user is who they say they are. Authorization is used to verify that a user has permission to do something.
+
+## Starter Application
+
+In this tutorial, you'll work on authentication middleware for an existing API built with Flask and PyMongo. The API is a book library API using which users can create books and upload cover images for the books and relevant data. PyMongo is used to connect to the mongo database. You'll use the PyJWT library to generate and verify JWT tokens. 
+
+To get started, clone the repository and set up the application by running the following commands:
+
+```bash 
+git clone https://github.com/Babatunde13/login-radius-flask.git # Clone the repository
+cd loginRadius-flask-auth # change directory
+git checkout start # check out the start branch
+virtualenv venv # create virtual environment
```

Comment:
@Babatunde13 I guess, you need to update when the reader should use `virtualenv` (Python 2) or `venv` (Python 3) based on https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/